HDDS-5637. Add documentation for Recon Namespace REST APIs (#2566)
apache/ozone: hadoop-hdds/docs/content/interface/ReconApi.md

Returns the UnhealthyContainerMetadata objects for the containers in the givenstate.
Possible unhealthy container states are `MISSING`, `MIS_REPLICATED`,`UNDER_REPLICATED`, `OVER_REPLICATED`.
The response structure is same as `/containers/unhealthy`.

## Namespace Metadata

### GET /api/v1/namespace/summary

**Parameters**

* path

   The path request in string without any protocol prefix.

**Returns**

Returns a basic summary of the path, including entity type and aggregate count 
of objects under the path.

`status` is `OK` if path exists, `PATH_NOT_FOUND` otherwise.

> 

Reviewer: ```suggestion
Example: /api/v1/namespace/summary?path=/

```

```json
     {
     	"status": OK,
     	"type": ROOT,
     	"numVolume": 10,
        "numBucket": 100,
        "numDir": 1000,

>         "numKey": 10000

Reviewer: Indentation

     }
```

If any `num` field is `-1`, the path request is not applicable to such an entity type.

### GET /api/v1/namespace/du

**Parameters**

* path

  The path request in string without any protocol prefix.

* files (optional)

  A boolean with a default value of `false`. If set to `true`, compute disk usage for keys 
  directly under the path.

* replica (optional)

  A boolean with a default value of `false`. If set to `true`, compute disk usage with replicated
size of keys.

**Returns**

Returns disk usage of all sub-paths under the path. Normalizes `path` fields, returns
total size of keys directly under the path as `sizeDirectKey`, and returns 
`size`/`sizeWithReplica` in number of bytes. 

> `status` is `OK` if path exists, 
> `PATH_NOT_FOUND` otherwise.

Reviewer: ```suggestion
`status` is `OK` if path exists, `PATH_NOT_FOUND` otherwise.
```

> 

Reviewer: Let's put the example query path & params that gives the result below here as well.

```json
     {
     	"status": OK,
        "path": "/vol1/bucket1",
        "size": 100000,
        "sizeWithReplica": 300000,
        "subPathCount": 4,
        "subPaths": [
          {
            "path": "/vol1/bucket1/dir1-1",
            "size": 30000,
            "sizeWithReplica": 90000,
            "isKey": false
          },
          {
            "path": "/vol1/bucket1/dir1-2",
            "size": 30000,
            "sizeWithReplica": 90000,
            "isKey": false
          },
          {
            "path": "/vol1/bucket1/dir1-3",
            "size": 30000,
            "sizeWithReplica": 90000,
            "isKey": false
          },
          {
            "path": "/vol1/bucket1/key1-1",
            "size": 10000,
            "sizeWithReplica": 30000,
            "isKey": true
          }
        ],

>         "sizeDirectKey": 10000

Reviewer: Indentation

     }
```
If `files` is set to `false`, sub-path `/vol1/bucket1/key1-1` is omitted.
If `replica` is set to `false`, `sizeWithReplica` returns `-1`. If the path's entity type
cannot have direct keys (Root, Volume), `sizeDirectKey` returns `-1`.

### GET /api/v1/namespace/quota

**Parameters**

* path

  The path request in string without any protocol prefix.

**Returns**

Returns the quota allowed and used under the path. Only volumes and buckets
have quota. Other types are not applicable to the quota request.

`status` is `OK` if the request is valid, `PATH_NOT_FOUND` if path doesn't exist,

> `TYPE_NOT_APPLIABLE` if path exists, but the path's entity type is not applicable

Reviewer: Typo?
```suggestion
`TYPE_NOT_APPLICABLE` if path exists, but the path's entity type is not applicable
```

to the request.

> 

Reviewer: Same here.

```json
     {
     	"status": OK,

>         "allowed": 200000,
>         "used": 160000

Reviewer: Why indent 3 spaces for `allowed` and `used` instead of 2?

```suggestion
       "allowed": 200000,
       "used": 160000
```

     }
```

If quota is not set, `allowed` returns `-1`. More on [Quota in Ozone]
(https://ci-hadoop.apache.org/view/Hadoop%20Ozone/job/ozone-doc-master/lastSuccessfulBuild/artifact/hadoop-hdds/docs/public/feature/quota.html)

### GET /api/v1/namespace/dist

**Parameters**

* path

  The path request in string without any protocol prefix.

**Returns**

Returns the file size distribution of all keys under the path.

`status` is `OK` if the request is valid, `PATH_NOT_FOUND` if path doesn't exist,

> `TYPE_NOT_APPLIABLE` if path exists, but the path is a key, which does not have

Reviewer: Another typo?
```suggestion
`TYPE_NOT_APPLICABLE` if path exists, but the path is a key, which does not have
```

a file size distribution.

> 

Reviewer: Same

```json
     {
     	"status": OK,
        "dist": [
          0,
          0,
          10,
          20,
          0,
          30,
          0,
          100,
          ...

>         ]

Reviewer: Same here.

     }
```

> Recon keeps track of all keys with size from `1KB` to `1PB`. For keys smaller than `1KB`,
> map to the first bin (index); for keys larger than `1PB`, map to the last bin (index).
> 
> Each index of `dist` is mapped to a file size range (e.g. `1MB` - `2MB`).

Reviewer: ```suggestion

Recon keeps track of all keys with size from `1 KB` to `1 PB`. For keys smaller than `1 KB`,
map to the first bin (index); for keys larger than `1 PB`, map to the last bin (index).

Each index of `dist` is mapped to a file size range (e.g. `1 MB` - `2 MB`).
```

Returns the UnhealthyContainerMetadata objects for the containers in the givenstate.
Possible unhealthy container states are `MISSING`, `MIS_REPLICATED`,`UNDER_REPLICATED`, `OVER_REPLICATED`.
The response structure is same as `/containers/unhealthy`.

## Namespace Metadata

### GET /api/v1/namespace/summary

**Parameters**

* path

   The path request in string without any protocol prefix.

**Returns**

Returns a basic summary of the path, including entity type and aggregate count 
of objects under the path.

`status` is `OK` if path exists, `PATH_NOT_FOUND` otherwise.

> Example: /api/v1/namespace/summary?path=/

Reviewer 2: Please include another example for a non root path.

Author: Thanks @kerneltime for taking a look! Do we need examples of non-root path for all endpoints?

```json
    {
      "status": OK,
      "type": ROOT,
      "numVolume": 10,
      "numBucket": 100,
      "numDir": 1000,
      "numKey": 10000
    }
```

If any `num` field is `-1`, the path request is not applicable to such an entity type.

### GET /api/v1/namespace/du

**Parameters**

* path

  The path request in string without any protocol prefix.

* files (optional)

  A boolean with a default value of `false`. If set to `true`, compute disk usage for keys 

>   directly under the path.

Reviewer 2: A boolean with a default value of `false`. If set to `true`, _computes_ disk usage for keys 
   under the path.
